hail/python/hailtop/batch/functions.py

```diff
@@ -0,0 +1,111 @@
+import math
+
+from typing import List
+
+from ..utils.utils import grouped
+from .utils import BatchException
+from .batch import Batch
+from .resource import ResourceGroup, ResourceFile
+
+
+def concatenate(b: Batch, files: List[ResourceFile], branching_factor: int = 100):
+    """
+    Concatenate files using tree aggregation.
+
+    Notes
+    -----
+    Runs the `cat` command aggregating files using the `branching_factor`.
+
+    Examples
+    --------
+    Create and execute a batch that concatenates output files:
+
+    >>> b = Batch()
+    >>> j1 = b.new_job()
+    >>> j1.command(f'touch {j1.ofile}')
+    >>> j2 = b.new_job()
+    >>> j2.command(f'touch {j2.ofile}')
+    >>> j3 = b.new_job()
+    >>> j3.command(f'touch {j3.ofile}')
+    >>> files = [j1.ofile, j2.ofile, j3.ofile]
+    >>> ofile = concatenate(b, files, branching_factor=2)
+    >>> b.run()
+
+    Parameters
+    ----------
+    b:
+        Batch to add concatenation jobs to.
+    files:
+        List of files to concatenate.
+    branching_factor:
+        Grouping factor when concatenating files.
+
+    Returns
+    -------
+    Concatenated output file.
+    """
+
+    def _concatenate(b, name, xs):
+        j = b.new_job(name=name)
+        j.command(f'cat {" ".join(xs)} > {j.ofile}')
+        return j.ofile
+
+    if len(files) == 0:
+        raise BatchException(f'Must have at least one file to concatenate.')
```

Comment:
Hmm. So, I understand the desire to warn the user. Concatenating only one file also seems warning worthy?

In the spirit of flexible, generic tools, it seems like `len(files) == 0` should be `j.comand(f'touch {j.ofile}')`, what do you think of that? Maybe I'm pushing things too far towards a software-engineering-mindset.

Reply:
I don't like this. It should just be whatever the cat command accepts as valid inputs.